flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1640,13 +1640,102 @@ struct InsertValueOpConversion
 
 /// InsertOnRange inserts a value into a sequence over a range of offsets.
 struct InsertOnRangeOpConversion
-    : public FIROpAndTypeConversion<fir::InsertOnRangeOp> {
+    : public FIROpAndTypeConversion<fir::InsertOnRangeOp>,
+      public ValueOpCommon {
   using FIROpAndTypeConversion::FIROpAndTypeConversion;
 
+  // Increments an array of subscripts in a row major fasion.
+  void incrementLoop(const SmallVector<uint64_t, 8> &dims,
+                     SmallVector<uint64_t, 8> &subscripts) const {
+    for (size_t i = dims.size(); i > 0; --i) {
+      if (subscripts[i - 1]++ < dims[i - 1] - 1) {
+        return;
+      }
+      subscripts[i - 1] = 0;
+    }
+  }
+
   mlir::LogicalResult
   doRewrite(fir::InsertOnRangeOp range, mlir::Type ty, OperandTy operands,
             mlir::ConversionPatternRewriter &rewriter) const override {
-    TODO();
+    assert(fir::allConstants(operands.drop_front(2)));
+    llvm::errs() << "InsertOnRange\n";
+
+    SmallVector<mlir::Attribute, 8> lowerBound;
+    SmallVector<mlir::Attribute, 8> upperBound;
+    SmallVector<uint64_t, 8> dims;
+    // mlir::LLVM::LLVMType type =
+    auto type = operands[0].getType().dyn_cast<mlir::LLVM::LLVMType>();
+
+    // Iterativly extract the array dimensions from it's type.
+    while (type.isArrayTy()) {
+      dims.push_back(type.getArrayNumElements());
+      type = type.getArrayElementType();
+    }
+
+    // Unzip the upper and lower bound subscripts.
+    for (std::size_t i = 2; i < operands.size(); i += 2) {
+      lowerBound.push_back(ExtractValueOpConversion::getValue(operands[i]));
+      upperBound.push_back(ExtractValueOpConversion::getValue(operands[i + 1]));
+    }
+
+    // ExtractValueOpConversion::toRowMajor(
+    //    lowerBound, lowerTy().unwrap(operands[0].getType()));
+
+    SmallVector<uint64_t, 8> lBounds;
+    SmallVector<uint64_t, 8> uBounds;
+
+    // Extract the integer value from the attribute bounds and convert to row
+    // major format.
+    for (size_t i = lowerBound.size(); i > 0; --i) {
+      lBounds.push_back(lowerBound[i - 1].cast<IntegerAttr>().getInt());
+      uBounds.push_back(upperBound[i - 1].cast<IntegerAttr>().getInt());
+    }
+
+    auto subscripts(lBounds);
+    auto loc = range.getLoc();
+    mlir::Value lastOp = operands[0];
+    mlir::Value insertVal = operands[1];
+
+    while (true) {
+      // Check to see if the current subscript has reached the upper bound.
+      bool equal = true;
+      for (size_t i = 0; i < subscripts.size(); ++i) {
+        if (subscripts[i] != uBounds[i]) {
+          equal = false;
+          break;
+        }
+      }
+      if (equal) {
+        break;
+      }
+
+      // Convert uint64_t's to Attribute's.
+      SmallVector<mlir::Attribute, 8> subscriptAttrs;
+      for (const auto &subscript : subscripts) {
+        subscriptAttrs.push_back(
+            IntegerAttr::get(rewriter.getI64Type(), subscript));
+      }
```

Comment:
No `{}` for this for loop with LLVM style.

Reply:
My bad, its been a little while since I read through the LLVM code style.